File: tuhubread/lib/models/voucher_save.model.dart
import 'package:json_annotation/json_annotation.dart';

import 'voucher.model.dart';

part 'voucher_save.model.g.dart';

@JsonSerializable(fieldRename: FieldRename.snake)
class VoucherSaveModel {
  @JsonKey(name: '_id')
  final String id;

  @JsonKey(name: 'voucher_id')
  final VoucherModel? voucher;
  final String voucherCode; // snapshot mã lúc save
  final DateTime expiresAt; // snapshot end_date lúc save
  final String status; // saved / used / expired
  final DateTime savedAt;
  final DateTime? usedAt;

  VoucherSaveModel({
    required this.id,
    this.voucher,
    required this.voucherCode,
    required this.expiresAt,
    required this.status,
    required this.savedAt,
    this.usedAt,
  });

  bool get isUsed => status == 'used';
  bool get isExpired =>
      status == 'expired' || expiresAt.isBefore(DateTime.now());
  bool get isAvailable => status == 'saved' && !isExpired;

  factory VoucherSaveModel.fromJson(Map<String, dynamic> json) =>
      _$VoucherSaveModelFromJson(json);

  Map<String, dynamic> toJson() => _$VoucherSaveModelToJson(this);
}

File: be/models/voucherSave.model.js
const db = require("../configs/db");

// Ghi nhận user đã save (claim) voucher.
// Unique index (voucher_id + user_id): 1 user chỉ save 1 voucher 1 lần.
const voucherSaveSchema = new db.mongoose.Schema(
  {
    voucher_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "voucherModel",
      required: true,
    },
    user_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "userModel",
      required: true,
    },
    voucher_code: { type: String, required: true }, // snapshot mã lúc save
    expires_at: { type: Date, required: true },     // snapshot end_date lúc save
    status: {
      type: String,
      required: true,
      enum: ["saved", "used", "expired"],
      default: "saved",
    },
    saved_at: { type: Date, required: true, default: Date.now },
    used_at: { type: Date, default: null },
    deleted_at: { type: Date, default: null },
  },
  { collection: "voucher_saves", timestamps: true },
);

// Đảm bảo mỗi user chỉ save mỗi voucher 1 lần
voucherSaveSchema.index({ voucher_id: 1, user_id: 1 }, { unique: true });

let voucherSaveModel = db.mongoose.model("voucherSaveModel", voucherSaveSchema);
module.exports = { voucherSaveModel };
